### ISSUE TITLE
CLI utf error

### ISSUE THREAD

**Issue by casperdcl** [tqdm/tqdm#663]
```python
python2 -c 'print("\xff")' | python3 -m tqdm # error
python2 -c 'print("\xff")' | python2 -m tqdm # works
python2 -c 'print(u"\xff")' | python3 -m tqdm # error
python2 -c 'print(u"\xff")' | python2 -m tqdm # error
```

**Comment by casperdcl** [comment]
related: #314 #598

**Comment by casperdcl** [comment]
@altendky maybe want to take a look? Hopefully I'll get to #598 soon

**Comment by altendky** [comment]
@casperdcl it might help to have the actual full prompts/commands/output. :]  Also, include the command `locale`.

**Comment by altendky** [comment]
Oh, and also run without the pipes.  Just the prints.

**Comment by casperdcl** [comment]
```
$ locale
LANG=en_GB.UTF-8
LANGUAGE=en_GB:en
LC_CTYPE="en_GB.UTF-8"
LC_NUMERIC="en_GB.UTF-8"
LC_TIME="en_GB.UTF-8"
LC_COLLATE="en_GB.UTF-8"
LC_MONETARY="en_GB.UTF-8"
LC_MESSAGES="en_GB.UTF-8"
LC_PAPER="en_GB.UTF-8"
LC_NAME="en_GB.UTF-8"
LC_ADDRESS="en_GB.UTF-8"
LC_TELEPHONE="en_GB.UTF-8"
LC_MEASUREMENT="en_GB.UTF-8"
LC_IDENTIFICATION="en_GB.UTF-8"
LC_ALL=en_GB.UTF-8

$ python2 -c 'print("\xff")' | python3 -m tqdm
0it [00:00, ?it/s]Traceback (most recent call last):
  File "/usr/lib/python3.5/runpy.py", line 184, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.5/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/cc16/gits/tqdm/tqdm/__main__.py", line 2, in <module>
    main()
  File "/home/cc16/gits/tqdm/tqdm/_main.py", line 217, in main
    for i in tqdm(sys.stdin, **tqdm_args):
  File "/home/cc16/gits/tqdm/tqdm/_tqdm.py", line 1002, in __iter__
    for obj in iterable:
  File "/usr/lib/python3.5/codecs.py", line 321, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte
$ python2 -c 'print("\xff")'
�

$ python2 -c 'print("\xff")' | python2 -m tqdm
0it [00:00, ?it/s]�
1it [00:00, 1286.99it/s]


$ python2 -c 'print(u"\xff")' | python3 -m tqdm
Traceback (most recent call last):
  File "<string>", line 1, in <module>
UnicodeEncodeError: 'ascii' codec can't encode character u'\xff' in position 0: ordinal not in range(128)
0it [00:00, ?it/s]
$ python2 -c 'print(u"\xff")'
ÿ

$ python2 -c 'print(u"\xff")' | python2 -m tqdm
Traceback (most recent call last):
  File "<string>", line 1, in <module>
UnicodeEncodeError: 'ascii' codec can't encode character u'\xff' in position 0: ordinal not in range(128)
0it [00:00, ?it/s]
$ python2 -c 'print(u"\xff")'
ÿ
```

**Comment by altendky** [comment]
> ```
> $ python2 -c 'print("\xff")' | python3 -m tqdm
> 0it [00:00, ?it/s]Traceback (most recent call last):
>   File "/usr/lib/python3.5/runpy.py", line 184, in _run_module_as_main
>     "__main__", mod_spec)
>   File "/usr/lib/python3.5/runpy.py", line 85, in _run_code
>     exec(code, run_globals)
>   File "/home/cc16/gits/tqdm/tqdm/__main__.py", line 2, in <module>
>     main()
>   File "/home/cc16/gits/tqdm/tqdm/_main.py", line 217, in main
>     for i in tqdm(sys.stdin, **tqdm_args):
>   File "/home/cc16/gits/tqdm/tqdm/_tqdm.py", line 1002, in __iter__
>     for obj in iterable:
>   File "/usr/lib/python3.5/codecs.py", line 321, in decode
>     (result, consumed) = self._buffer_decode(data, self.errors, final)
> UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte
> ```

This makes sense to me.  As it says, `\xff` isn't valid UTF-8, at least not as a first byte.

> ```
> $ python2 -c 'print("\xff")'
> �
> ```

This seems to agree with above about it being invalid.

> ```
> $ python2 -c 'print("\xff")' | python2 -m tqdm
> 0it [00:00, ?it/s]�
> 1it [00:00, 1286.99it/s]
> ```

Unlike Python 3, Python 2 doesn't decode the bytes coming in, it happily throws around bytes and the terminal is similarly confused when it gets `\xff`.

> ```
> $ python2 -c 'print(u"\xff")' | python3 -m tqdm
> Traceback (most recent call last):
>   File "<string>", line 1, in <module>
> UnicodeEncodeError: 'ascii' codec can't encode character u'\xff' in position 0: ordinal not in range(128)
> 0it [00:00, ?it/s]
> $ python2 -c 'print(u"\xff")'
> ÿ
> ```

Given the local I would have expected it to use the UTF-8 encoding, but I forget some details here.  Anyways, on the Python 2 side here you've got a Unicode object which is being implicitly encoded but the character that was forced in there isn't valid ASCII so it fails.

> ```
> $ python2 -c 'print(u"\xff")' | python2 -m tqdm
> Traceback (most recent call last):
>   File "<string>", line 1, in <module>
> UnicodeEncodeError: 'ascii' codec can't encode character u'\xff' in position 0: ordinal not in range(128)
> 0it [00:00, ?it/s]
> $ python2 -c 'print(u"\xff")'
> ÿ
> ```

Same as above.

tl;dr Don't expect random things to be able to be encoded and decoded, only some encodings consider every byte sequence to be valid and only some encodings are able to encode all Unicode characters.

**Comment by altendky** [comment]
```
altendky@lt:~/grid-tied.green$ python2 -c 'import sys; sys.stderr.write(repr(sys.stdout.encoding) + "\n")'
'UTF-8'
altendky@lt:~/grid-tied.green$ python2 -c 'import sys; sys.stderr.write(repr(sys.stdout.encoding) + "\n")' | cat
None
```

Yes, the pipe does affect how Python handles the encoding.

**Comment by casperdcl** [comment]
Ah I wasn't explicit enough about why this is an issue.

> Unlike Python 3, Python 2 doesn't decode the bytes coming in, it happily throws around bytes and the terminal is similarly confused when it gets \xff.

That's the point, `tqdm` should be usable as a pipe on the CLI without encoding/decoding anything and "happily throw around bytes", i.e.:

```
tar cf - docs/ | tqdm --bytes > backup.tar
```

should *never* error.

> Yes, the pipe does affect how Python handles the encoding.

... there are no words for this evil

**Comment by altendky** [comment]
```
altendky@lt:~$ python2 -c 'print("\xff")' | python3 -c 'import sys; print(sys.stdin.read())'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/altendky/.pyenv/versions/3.7.2/lib/python3.7/codecs.py", line 322, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte
```

```
altendky@lt:~$ python2 -c 'print("\xff")' | python3 -c 'import sys; print(sys.stdin.buffer.read())'
b'\xff\n'
```

```
altendky@lt:~$ python2 -c 'print("\xff")' | python3 -c 'import sys; sys.stderr.buffer.write(sys.stdin.buffer.read())'
�
```

**Comment by altendky** [comment]
You probably want to iterate... But something like that seems at least closer. You have to figure out when to do that though.  But maybe that's what `--bytes` is for.

**Comment by casperdcl** [comment]
No, the iteration handling code I'd written still works fine. 7550d31 seems to fix the issue. Would like a unit test to prevent regression but can't figure out how.

**Comment by casperdcl** [comment]
test somewhere here:

https://github.com/tqdm/tqdm/blob/devel/tqdm/tests/tests_main.py#L50

**Comment by mrrqzhang** [comment]
open(encoding="UTF-8", errors="ignore")

**Comment by casperdcl** [comment]
@mrrqzhang is that a patch which fixes something for you?
